ipa-core/src/hpke/registry.rs

```diff
@@ -102,13 +107,23 @@ impl KeyRegistry<KeyPair> {
             keys: keys.into_boxed_slice(),
         }
     }
+}
 
+impl PrivateKeyRegistry for KeyRegistry<KeyPair> {
     #[must_use]
-    pub(super) fn private_key(&self, key_id: KeyIdentifier) -> Option<&IpaPrivateKey> {
+    fn private_key(&self, key_id: KeyIdentifier) -> Option<&IpaPrivateKey> {
         self.key(key_id).map(|v| &v.sk)
     }
 }
 
+impl PrivateKeyRegistry for KeyRegistry<IpaPrivateKey> {
+    #[must_use]
+    fn private_key(&self, key_id: KeyIdentifier) -> Option<&IpaPrivateKey> {
+        self.key(key_id)
+    }
+}
+
+
```

Comment:
This is giving me the following compile error (thought it seems to be the same approach as `PublicKeyRegistry` just below):
```
error[E0119]: conflicting implementations of trait `PrivateKeyRegistry` for type `KeyRegistry<hpke::registry::KeyPair>`
   --> ipa-core/src/hpke/registry.rs:119:1
    |
112 | impl PrivateKeyRegistry for KeyRegistry<KeyPair> {
    | ------------------------------------------------ first implementation here
...
119 | impl PrivateKeyRegistry for KeyRegistry<IpaPrivateKey> {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `KeyRegistry<hpke::registry::KeyPair>`
    ```

Reply:
ah, yea this is the limitation of Rust trait system. You can't implement `Foo` for `Bar<X>` and `Bar<Y>` - it is considered to be the same definition. I think you may want to define a `PrivateKeyContainer` trait and then implement it for `KeyPair` and `IpaPrivateKey`. then you can implement `PrivateKeyRegistry` for any `K: PrivateKeyContainer`

```rust
impl <K: PrivateKeyContainer> PrivateKeyRegistry for KeyRegistry<K> { .. }
```

Reply:
ok, thanks. (I'm still confused why we can implement `PublicKeyRegistry` for both `KeyRegistry<KeyPair>` and `KeyRegistry<PublicKeyOnly>` just below. That seems to do exactly what this is complaining about.)

Reply:
ah, maybe because of this

```rust
// This newtype is necessary because IpaPublicKey is an associated type from another crate (hpke).
// The coherence rules prohibit us from implementing `PublicKeyRegistry` both for our concrete type
// `KeyPair` and for `IpaPublicKey`, because the impls would overlap if hpke chose to define
// `IpaPublicKey` to be the same as `KeyPair`.
pub struct PublicKeyOnly(pub IpaPublicKey);
```

My statement about `Bar<X>` and `Bar<Y>` was wrong, I confused generics vs associated types.